src/sdl2/render.rs

```diff
@@ -125,40 +178,233 @@ impl RendererInfo {
             flags: info.flags,
             texture_formats: texture_formats,
             max_texture_width: info.max_texture_width as u32,
-            max_texture_height: info.max_texture_height as u32
+            max_texture_height: info.max_texture_height as u32,
         }
     }
 }
 
-pub enum RendererParent<'a> {
-    Surface(Surface<'a>),
-    Window(Window)
-}
-
-/// 2D rendering context
-pub struct Renderer<'a> {
+/// Holds a `SDL_Renderer`
+///
+/// When the `RendererContext` is dropped, it destroys the `SDL_Renderer`
+pub struct RendererContext {
     raw: *mut ll::SDL_Renderer,
-    parent: Option<RendererParent<'a>>,
-    is_alive: Rc<UnsafeCell<bool>>
 }
 
-impl<'a> Drop for Renderer<'a> {
+impl Drop for RendererContext {
     fn drop(&mut self) {
         unsafe {
-            *self.is_alive.get() = false;
             ll::SDL_DestroyRenderer(self.raw);
         };
     }
 }
 
+impl RendererContext {
+    /// Gets information about the rendering context.
+    pub fn info(&self) -> RendererInfo {
+        unsafe {
+            let mut renderer_info_raw = mem::uninitialized();
+            if ll::SDL_GetRendererInfo(self.raw, &mut renderer_info_raw) != 0 {
+                // Should only fail on an invalid renderer
+                panic!();
+            } else {
+                RendererInfo::from_ll(&renderer_info_raw)
+            }
+        }
+    }
+
+    /// Gets the raw pointer to the SDL_Renderer
+    pub unsafe fn raw(&self) -> *mut ll::SDL_Renderer {
+        self.raw
+    }
+
+    pub unsafe fn from_ll(raw: *mut ll::SDL_Renderer) -> Self {
+        RendererContext { raw: raw }
+    }
+
+    unsafe fn set_raw_target(&self, raw_texture: *mut ll::SDL_Texture) -> Result<(), SdlError> {
+        if ll::SDL_SetRenderTarget(self.raw, raw_texture) == 0 {
+            Ok(())
+        } else {
+            Err(SdlError(get_error()))
+        }
+    }
+}
+
+impl<T> Deref for Canvas<T> {
+    type Target = RendererContext;
+
+    fn deref(&self) -> &RendererContext {
+        self.context.as_ref()
+    }
+}
+
+/// 2D Renderer
+///
+/// Manages the target on which we draw onto (e.g., `Window`, `Surface`, or a `Texture`)
+/// and the `RendererContext`
+///
+/// Its context may be shared with the `TextureCreator`.
+///
+/// The context will not be dropped until all references of it are out of scope.
+pub struct Canvas<T> {
```

Comment:
Unfortunately it was not as easy as calling mem::drop since that takes a `self` and our drop is only a `&mut self`. 

I had to restructure the code a bit (actually put it back to how it was friday). It is a bit unfortunate because it makes the Canvas have more generics but the type aliases should help with that.